### DISCUSSION TITLE
Params  and headers - None value (and comparison with requests)

### DISCUSSION THREAD

**Discussion by delicb** [encode/httpx#3040] [General]
I am working on migrating an HTTP library from `requests` to `httpx`. One of the differences I noticed (which matter for me in final implementation) is treatment of `None` values in `params` and `headers`. Requests just does not send params if they have `None` value, while `httpx` sends query parameter but without value and raises exceptions for headers. 

What requests does seems more intuitive to me. Is this deliberate design decision and is it subject to change? I can handle it on my end, but it seems like something that should be part of `httpx`.

For headers `requests` does not send `None` values, while `httpx` fails with `AttributeError: 'NoneType' object has no attribute 'encode'`. IMO - at very least it should not fail. 

Example:

```python
import httpx
import requests
import pytest

def test_compare_httpx_requests_params():
    import httpx
    import requests
    params = {"exists": "yes", "multi": [1, 2], "not-exists": None}
    req_res = requests.get("https://httpbin.org/get", params=params).json()["args"]
    httpx_res = httpx.get("https://httpbin.org/get", params=params).json()["args"]
    print(req_res)  # prints: {'exists': 'yes', 'multi': ['1', '2']}
    print(httpx_res)  # {'exists': 'yes', 'multi': ['1', '2'], 'not-exists': ''}


def test_compare_httpx_requests_headers():
    headers = {"X-H1": "yes", "X-H2": None}
    req_res = requests.get("https://httpbin.org/get", headers=headers).json()["headers"]
    with pytest.raises(AttributeError):
        httpx.get("https://httpbin.org/get", headers=headers).json()  # raises AttributeError
    print(req_res)  # prints headers including only X-H1
```

I think I saw discussion about overwriting client params in concrete request - this might be a solution to that as well (send `None` for concrete value). 

What do you think?

**Comment by Fury1** [comment]
I just ran into this today myself switching over from requests with a "params" argument in GET requests.

I did find this -> [https://www.python-httpx.org/compatibility/#query-parameters](https://www.python-httpx.org/compatibility/#query-parameters)

Seems like it is a documented design choice as httpx aims to be mostly compatible but not a drop in replacement.

I just made a payload filter to handle things:

```
f_payload = {k: v for k, v in payload.items() if v is not None}
```
